src/raw/client.rs

```diff
@@ -818,6 +800,58 @@ impl<PdC: PdClient> Client<PdC> {
         Ok(result)
     }
 
+    async fn retryable_scan(
+        &self,
+        mut scan_args: ScanInnerArgs,
+    ) -> Result<(Option<RawScanResponse>, Key)> {
+        let start_key = scan_args.start_key;
+
+        loop {
+            let region = self.rpc.clone().region_for_key(&start_key).await?;
+            let store = self.rpc.clone().store_for_id(region.id()).await?;
+            let request = new_raw_scan_request(
+                scan_args.range.clone(),
```

Comment:
I agree with @pingyu and IIUC the problem is that `scan_args.range` never changes across iterations. 

`scan_args.start_key` is totally fine; it moves forward as the scan makes progress. But since `scan_args.range` doesn't change, it's possible for `retryable_scan` to have `scan_args.range=(key000, key100) scan_args.start_key=key050`. In this case, if we encounter a region merge, the raw scan request constructed here will scan the new region from key000 again, getting duplicate KVs. 

I think one way to fix it is to gradually shrink the current range in the loop of `scan_inner`, similar to what the old code did:
```
cur_range = BoundRange::new(
    std::ops::Bound::Included(region_store.region_with_leader.range().1),
    cur_range.to,
);
```

Reply:
Yes, now I see it. The scan request is being constructed on the range but the range itself is not being changed for every iteration. So essentially, mutating the range with the updated start_key will fix the issue. Thanks @hbisheng and @pingyu for being so patient with the explanation. Adding the fix.